File: jvm/aula-10-27/src/main/java/pc/li52d/monitors/SemaphoreFifoSN.java
package pc.li52d.monitors;

import pc.li52d.monitors.utils.NodeList;
import pc.li52d.monitors.utils.TimeoutHolder;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

public class SemaphoreFifoSN {
    /**
     *  this class instances represent
     *  a pending acquire request
     */
    private static class Request {
        public final int units;
        public final Condition condition;
        public boolean done;

        public Request(int units, Condition condition) {
            this.units = units;
            this.done = false;
            this.condition = condition;
        }
    }

    private int units;
    private final NodeList<Request> requests;
    private final ReentrantLock monitor;

    public SemaphoreFifoSN(int initial) {
        if (initial > 0)
            units = initial;
        monitor = new ReentrantLock();
        requests = new NodeList<>();
    }

    /**
     * An auxiliary method that tries
     * to process all possible pending requests
     */
    private void notifyWaiters() {
        while (requests.size() > 0 && units >= requests.first().units) {

            Request r = requests.removeFirst();
            units-= r.units;
            r.done = true;
            r.condition.signal();

        }
    }

    public boolean acquire(int n, long millis)
        throws InterruptedException {
        monitor.lock();
        try {
            // fast path
            if (units >= n && requests.empty()) {
                units -= n;
                return true;
            }
            if (millis == 0) return false;

            // prepare wait
            var req = new Request(n, monitor.newCondition());
            var node = requests.addLast(req);
            TimeoutHolder th = new TimeoutHolder(millis);

            // do wait
            do {
                try {
                    // non interruption path
                    req.condition.await(th.remaining(),
                        TimeUnit.MILLISECONDS);
                    if (req.done) return true;
                    if (th.timeout()) {
                        requests.remove(node);
                        // note that we must notify
                        // in this case, since the
                        // timeouted request
                        // can be at the front of the list
                        notifyWaiters();
                        return false;
                    }
                }
                catch(InterruptedException e) {
                    // note that the request
                    // can be already accepted
                    // so in the case we return success
                    // but redoing a thread interrupt request
                    if (req.done) {
                        Thread.currentThread().interrupt();
                        return true;
                    }
                    // note that we must notify
                    // in this case, since the
                    // interrupted request
                    // can be at the front of the list
                    requests.remove(node);
                    notifyWaiters();
                    throw e;
                }
            }
            while(true);
        }
        finally {
            monitor.unlock();
        }

    }

    public void release(int n) {
        monitor.lock();
        try {
            units += n;
            notifyWaiters();
        }
        finally {
            monitor.unlock();
        }
    }
}

File: dotnet/Monitors/SemaphoreFifoSN.cs
using SynchUtils;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;

namespace Monitors
{
    public class SemaphoreFifoSN
    {
        /**
         *  this class instances represent
         *  a pending acquire request
         */
        private class Request
        {
            public object condition;
            public readonly int units;
            public bool done;

            public Request(int units) {
                condition = new object();
                this.units = units;
                this.done = false;
            }
        }

        private int units;
        private readonly LinkedList<Request> requests;
        private readonly object monitor;


        public SemaphoreFifoSN(int initial) {
            if (initial > 0)
                units = initial;
            monitor = new object();

            requests = new LinkedList<Request>();
        }

        /**
         * An auxiliary method that tries
         * to process all possible pending requests
         */
        private void NotifyWaiters() {
           
            while (requests.Count > 0 && units >= requests.First.Value.units) {

                Request r = requests.First.Value;
                requests.RemoveFirst();
                units -= r.units;
                r.done = true;
                r.condition.Notify(monitor);
            }
            
        }

        public bool Acquire(int n, int millis) {
            lock (monitor) {

                // fast path
                if (units >= n && requests.Count == 0) {
                    units -= n;
                    return true;
                }
                if (millis == 0) return false;

                // prepare wait
                var req = new Request(n);
                var node = requests.AddLast(req);
                TimeoutHolder th = new TimeoutHolder(millis);

                // do wait
                do {
                    try {
                        // non interruption path
                        req.condition.Wait(monitor, th.Remaining);
                      
                        if (req.done) return true;
                        if (th.Timeout) {
                            requests.Remove(node);
                            // note that we must notify
                            // in this case, since the
                            // timeouted request
                            // can be at the front of the list
                            NotifyWaiters();
                            return false;
                        }
                    }
                    catch (ThreadInterruptedException e) {
                        // note that the request
                        // can be already accepted
                        // so in the case we return success
                        // but redoing a thread interrupt request
                        if (req.done) {
                            Thread.CurrentThread.Interrupt();
                            return true;
                        }
                        // note that we must notify
                        // in this case, since the
                        // interrupted request
                        // can be at the front of the list
                        requests.Remove(node);
                        NotifyWaiters();
                        throw e;
                    }
                }
                while (true);
            }
        }


        public void Release(int n) {
            lock (monitor) {
                units += n;
                NotifyWaiters();
            }
        }
    }
}
